### PR TITLE
Added DSTC8-Synthetic Probabilistic Soft Logic model.

### PR DESCRIPTION
**Pull request by cfpryor**
Created a DSTC8-synthetic Probabilistic Soft Logic (PSL) model with 6 basic structural rules and a constants file used for running the model.

PoC: jereliu

### CODE REVIEW

**Review by jereliu — APPROVED** [review]

File: experimental/language_structure/psl/psl_model_dstc_synthetic.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        losses = []
+
+        for rule_weight, rule_function in zip(self.rule_weights,
+                                              self.rule_functions):
+            losses.append(rule_weight * rule_function(**rule_kwargs))
+        return losses
+
+    def compute_loss(self, data: tf.Tensor, logits: tf.Tensor) -> float:
+        """Calculate the total loss for all PSL rules."""
+        return sum(self.compute_loss_per_rule(data, logits))
```
Comment (jereliu) [inline experimental/language_structure/psl/psl_model_dstc_synthetic.py:398]: Let's consolidate those functions by only keeping the two, one for computing loss per rule, the other computing the total loss by summing up loss per rule.

**Review by jereliu — CHANGES_REQUESTED** [review]
Thanks Connor! Please see comment from reviewer (mostly about consolidating loss calculation functions).

File: experimental/language_structure/psl/psl_model_dstc_synthetic.py (hunk 1 above)
Comment (cfpryor) [reply experimental/language_structure/psl/psl_model_dstc_synthetic.py]: Nice catch, that was an extra function not needed.

File: experimental/language_structure/psl/psl_model_dstc_synthetic.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
```
Comment (jereliu) [inline experimental/language_structure/psl/psl_model_dstc_synthetic.py:388]: Is it possible to fold this into `compute_total_loss`?

**Review by jereliu — CHANGES_REQUESTED** [review]

File: experimental/language_structure/psl/psl_model_dstc_synthetic.py (hunk 2 above)
Comment (cfpryor) [reply experimental/language_structure/psl/psl_model_dstc_synthetic.py:378]: I want to mirror the multiwoz synthetic psl functions, therefore instead of folding this into compute_total_loss, it makes sense to remove compute_total_loss.

Comment (jereliu) [reply experimental/language_structure/psl/psl_model_dstc_synthetic.py:378]: Thanks for the change!

**Review by jereliu — APPROVED** [review]

**Review by jereliu — APPROVED** [review]

**Comment by jereliu** [comment]
Closing as this pull request has been fulfilled by https://github.com/google/uncertainty-baselines/pull/764
